File: bongf/python/level1/failure_rate.py
## 이중포문을 돌면 시간초과 
def solution(N, stages):
    rate = {}
    for i in range(1, N+1):
        d = 0
        n = 0
        for j in range(len(stages)):
            if stages[j] >= i :
                d += 1
                if stages[j] == i:
                    n += 1
        if d==0 :
            rate[i] = 0
        else :   
            rate[i] = n / d
    
    return sorted(rate, key = lambda x: rate[x], reverse = True)

def solution_old(N, stages):
    fail_dic = {}
    d = len(stages)
    for stage in range(1, N+1):
        if d !=0:
            count = stages.count(stage)
            fail_dic[stage] = count/d
            d -= count
        else:
            fail_dic[stage] = 0 
    return sorted(fail_dic, reverse=True, key=lambda x: fail_dic[x])

File: programmers/java/src/level1/FailureRate.java
package level1;

import java.util.*;

public class FailureRate {
    public static void main(String[] args) {
        System.out.println(solution2(5, new int[]{2, 1, 2, 6, 2, 4, 3, 3}));
    }

    //이게 더 빨랐다
    public static Integer[] solution(int N, int[] stages) {
        int[] ing = new int[N+2];
        Map<Integer, Double> rate = new HashMap<>();
        for (int stage : stages) {
            ing[stage]++;
        }
        for (int i = 1; i <=N; i++) {
            int triedP = Arrays.stream(Arrays.copyOfRange(ing, i, ing.length)).sum();
            if( triedP == 0 ) {
                rate.put(i, 0.0);
            } else {
                rate.put(i, (double)ing[i] / triedP);
            }
        }
        return rate.entrySet().stream()
                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                .map(Map.Entry::getKey)
                .toArray(Integer[]::new);
    }

    public static Integer[] solution2(int N, int[] stages) {
        Map<Integer, Double> rate = new HashMap<>();
        int p = stages.length;
        for (int i = 1; i <=N; i++) {
            int finalI = i;
            long c = Arrays.stream(stages)
                    .filter(x-> x== finalI)
                    .count();
            if(p==0) {
                rate.put(i, 0.0);
            } else {
                rate.put(i, (double)c / (p));
                p -= c;
            }
        }

        System.out.println(rate);

        return rate.entrySet().stream()
                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                .map(Map.Entry::getKey)
                .toArray(Integer[]::new);
    }
}
